File: go-backend-interview-prep/04_interfaces_and_generics/03_empty_interface_and_type_assertions/main.py
# Python equivalent of Empty Interface and Type Assertions (compare with main.go)
# Go any ~ Python object (everything is an object)
# Go type assertion ~ Python isinstance()


def describe_item(v: object) -> None:
    """isinstance = Python's type assertion."""
    if v is None:
        print("  None")
    elif isinstance(v, str):
        print(f"  string: {v!r} (len={len(v)})")
    elif isinstance(v, int) and not isinstance(v, bool):
        # bool is a subclass of int in Python, so check bool first or exclude it
        print(f"  int: {v}")
    else:
        print(f"  other: {v} ({type(v).__name__})")


def main():
    # --- object accepts anything (like Go's any) ---
    print("--- object accepts anything ---")
    v: object

    v = 42
    print(f"int:    {v} (type: {type(v).__name__})")

    v = "hello"
    print(f"string: {v} (type: {type(v).__name__})")

    v = True
    print(f"bool:   {v} (type: {type(v).__name__})")

    v = [1, 2, 3]
    print(f"list:   {v} (type: {type(v).__name__})")

    # --- isinstance = safe type assertion (like comma-ok) ---
    print("\n--- isinstance (safe assertion) ---")
    v = "Go is great"

    if isinstance(v, str):
        print(f"string assertion: val={v!r} ok=True")

    if isinstance(v, int):
        print(f"int assertion: val={v} ok=True")
    else:
        print(f"int assertion: val=0 ok=False")

    # --- No panic equivalent ---
    # Python has no "single-value assertion that panics."
    # The closest is: cast without check -> AttributeError later.
    print("\n--- Direct use (no assertion needed in Python) ---")
    v = "safe string"
    print("direct use:", v)  # Python is dynamically typed -- just use it

    # --- Processing a list of mixed types ---
    print("\n--- Processing list[object] ---")
    items: list[object] = [42, "hello", 3.14, True, None]
    for item in items:
        describe_item(item)

    # --- None ---
    print("\n--- None ---")
    nil_any = None
    print(f"None: {nil_any} (is None: {nil_any is None})")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/04_interfaces_and_generics/03_empty_interface_and_type_assertions/main.go
package main

import "fmt"

func main() {
	// --- any accepts anything ---
	fmt.Println("--- any accepts anything ---")
	var v any

	v = 42
	fmt.Printf("int:    %v (type: %T)\n", v, v)

	v = "hello"
	fmt.Printf("string: %v (type: %T)\n", v, v)

	v = true
	fmt.Printf("bool:   %v (type: %T)\n", v, v)

	v = []int{1, 2, 3}
	fmt.Printf("slice:  %v (type: %T)\n", v, v)

	// --- Type assertion: comma-ok form (SAFE) ---
	fmt.Println("\n--- Comma-ok type assertion ---")
	v = "Go is great"

	s, ok := v.(string)
	fmt.Printf("string assertion: val=%q ok=%t\n", s, ok)

	n, ok := v.(int)
	fmt.Printf("int assertion:    val=%d ok=%t\n", n, ok) // ok=false, n=0

	// --- Type assertion: single-value form (DANGEROUS) ---
	fmt.Println("\n--- Single-value assertion ---")
	v = "safe string"
	s2 := v.(string) // works because v IS a string
	fmt.Println("direct assertion:", s2)

	// v.(int) would PANIC here because v holds a string
	// Uncomment to see: _ = v.(int)

	// --- Practical: processing a slice of any ---
	fmt.Println("\n--- Processing []any ---")
	items := []any{42, "hello", 3.14, true, nil}
	for _, item := range items {
		describeItem(item)
	}

	// --- Nil any ---
	fmt.Println("\n--- Nil any ---")
	var nilAny any
	fmt.Printf("nil any: %v (is nil: %t)\n", nilAny, nilAny == nil)
}

func describeItem(v any) {
	if v == nil {
		fmt.Println("  nil")
		return
	}
	if s, ok := v.(string); ok {
		fmt.Printf("  string: %q (len=%d)\n", s, len(s))
		return
	}
	if n, ok := v.(int); ok {
		fmt.Printf("  int: %d\n", n)
		return
	}
	fmt.Printf("  other: %v (%T)\n", v, v)
}
